packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +984,89 @@ export class Select
         this.updateListboxMaxHeightCssVariable();
     }
 
+    private isOptionHidden(option: ListOption): boolean {
+        if (option.hidden) {
+            return true;
+        }
+        return !this.filterMatchesText(option.text);
+    }
+
+    private filterMatchesText(text: string): boolean {
+        const filter = this.filter.toLowerCase();
+        const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
+        return diacriticInsensitiveStringNormalizer(text).includes(
+            normalizedFilter
+        );
+    }
+
     /**
      * Filter available options by text value.
      *
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
-
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
-                );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
+        if (!this.$fastController.isConnected) {
+            return;
         }
 
-        this.options.forEach(o => {
-            if (isNimbleListOption(o)) {
-                if (!this.filteredOptions.includes(o)) {
-                    o.visuallyHidden = true;
-                } else {
-                    o.visuallyHidden = false;
+        const filteredOptions: ListOption[] = [];
+
+        let lastVisibleOptionGroupIndex = -1;
```

Comment:
I don't think you need to track `lastVisibleOptionGroupIndex` because you can set `lastVisibleElement.showBottomSeparator = true` when you find the next visible thing.

That would not only let you get rid of this variable and the check after the loop, but I think that would let you change the for loop to be `for (const element of this.slottedOptions)` rather than index based (which requires non-null assertion when indexing).